docs/specs/fee-calculator.md

```diff
@@ -0,0 +1,405 @@
+
+- Read access to stored distributor, balance, and event data through FileManager
+- Write permissions for fee calculation results
+
+## 4. Data Flow
+
+### Input Flow
+
+All data access exclusively through FileManager functions:
+
+1. Use FileManager to read list of all tracked distributors
+2. For each distributor:
+   - Use FileManager to read balance history data
+   - Use FileManager to read distribution event data
+3. Validate data completeness and consistency
+
+### Process Flow
+
+1. Load all distributor addresses from FileManager
+2. For each distributor:
+   - Retrieve complete balance history
+   - Retrieve all distribution events
+   - Sort data chronologically
+   - For each day with balance data:
+     - Calculate balance change from previous day
+     - Sum distributions for that day
+     - Compute daily fee = balance change + distributions
+   - Store calculated fees in memory
+3. Generate comprehensive fee report
+4. Write results using FileManager
+
+### Output Flow
+
+- Fee calculation results written through FileManager
+- Single report file containing all distributor daily fees
+- Structured format supporting easy querying and analysis
+
+## 5. Public API
+
+### Main Entry Point
+
+```typescript
+interface FeeCalculator {
+  calculateFees(distributorAddress?: string): Promise<void>;
+}
+```
+
+### Method Signatures
+
+```typescript
+calculateFees(distributorAddress?: string): Promise<void>
+```
+
+- **Purpose**: Calculate daily fees for all distributors or a specific distributor
+- **Parameters**:
+  - `distributorAddress` (optional): If provided, only calculate fees for this distributor
+- **Returns**: Promise that resolves when fee calculations are complete
+- **Behavior**:
+  - Reads all necessary data through FileManager
+  - Validates data completeness before processing
+  - Calculates daily fees for each distributor
+  - Writes comprehensive report with results
+  - Throws error if critical data is missing or invalid
+
+## 6. Algorithm Details
+
+### Fee Calculation Process
+
+1. **Initialize**: Create FileManager instance
+2. **Load Distributors**: Read distributor list using FileManager
+3. **Filter Distributors**: If distributorAddress provided, filter to single distributor
+4. **For Each Distributor**:
+   - Load balance history from FileManager
+   - Load distribution events from FileManager
+   - Build chronological timeline of all dates with data
+   - Sort balance entries by date
+   - For each date with balance data:
+     - If first date: balance change = 0 (no previous balance)
+     - Otherwise: balance change = current balance - previous balance
+     - Find all distribution events for this date
+     - Sum distribution amounts: total distributions = sum of all values
+     - Calculate daily fee = balance change + total distributions
+     - Store result with date, balances, distributions, and fee
+5. **Generate Report**: Compile all distributor results into structured report
```

Comment:
I don't know if it's clear from the spec or not, but calculating these fees is reasonably cheap. So we'll just ignore whatever report is there already and just completely overwrite it when we do this. I think that's what the plan here is already. I just want it to be clear in the spec that we're not like doing any incremental processing or anything like that. We just load all the data we need, do the calculations, put it back in the file manager.